---
title: "Analysis of Sox17-expressing lineages - GFP dataset"
output:
  html_document:
    df_print: paged
---

This notebook is used to analyze a single-cell RNA sequencing dataset of Sox17-expressing lineages (also known as the GFP dataset in the manuscript). Sorted viable GFP+ cells at E9.0-9.5 from Sox17GFPCre heterozygous males breeding with WT females were sequenced at sequencing service core,VANTAGE(Vanderbilt Technologies for Advanced Genomics). To confirm the identity of sorted GFP+cells, a customized genome build was generated by adding in GFP sequence to 10x Genomics reference (References - 2020-A, mm10) followed by standard alignment, UMI counting and barcode filtering using CellRanger 6.0.2
.

***

This analysis is performed with Seurat4 package. If Seurat is not installed or if certain version of Seurat is required, refer to satijalab.org for installation instructions.

***

# Setting up

```{r setup environment, include=FALSE}
all_times <- list() 
knitr::knit_hooks$set(time_it = local({
  now <- NULL
  function(before, options) {
    if(before){
      # report the current time before each chunk
      now <<- Sys.time()
    } else {
      # calculate the time difference after a chunk
      res <- difftime(Sys.time(), now, units = "secs")
      # store the reported time it takes to run each chunk to all_time object
      all_times[[options$label]] <<- res
    }
  }
}))
knitr::opts_chunk$set(
  # to reformat all R source code in several aspects such as adding spaces around most operators, indenting the code properly and replacing the assignment operator = into <-
  tidy = TRUE,
  # to set cutoff for the width of output text
  tidy.opts = list (width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  time_it = TRUE
)

```

```{r initiate library}
library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)
```

```{r load data}
sox17gfp.data <- Read10X(data.dir = "data/190620/filtered_feature_bc_matrix/")
```

```{r setup Seurat object}
# Initialize the Seurat object with the raw non-normalized data
sox17gfp <- CreateSeuratObject(counts=sox17gfp.data, project = "gfp", min.cells = 3, min.features = 200)

# Check Seurat object
sox17gfp
```

# Pre-processing

## Quality control

```{r percentage mitochondrial gene}
sox17gfp[["percent.mt"]] <- PercentageFeatureSet(sox17gfp, pattern = "^mt-")
```

```{r visualize QC metrics - before filter}
VlnPlot(sox17gfp, features = c("nFeature_RNA","nCount_RNA", "percent.mt"), ncol = 3)

plot1_gfp <- FeatureScatter(sox17gfp, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2_gfp <- FeatureScatter(sox17gfp, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1_gfp + plot2_gfp
```

```{r filter low-qual cells}
# Filter out cells that have unique feature counts over 8,000 or less than 200 or have more than 5% mitochondrial counts
sox17gfp <- subset(x = sox17gfp, subset = nFeature_RNA > 200 & nFeature_RNA < 8000 & percent.mt < 5)
```

```{r visualize QC metrics - after filter}
#Figure_S2B
VlnPlot(sox17gfp, features = c("nFeature_RNA","nCount_RNA", "percent.mt"), ncol = 3)

plot3_gfp <- FeatureScatter(sox17gfp, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot4_gfp <- FeatureScatter(sox17gfp, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot3_gfp + plot4_gfp
```

## Normalization

```{r glmGamPoi}
#glmGamPoi package can also be used to substantially improves the speed of the learning procedure. It can be invoked by specifying method="glmGamPoi"

#if(!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
#BiocManager::install("glmGamPoi")
```

```{r sctransform}
sox17gfp <- SCTransform(sox17gfp, method = "glmGamPoi", vars.to.regress = "percent.mt", verbose = FALSE)
```

# Dimensional reduction

## Linear dimentional reduction PCA

```{r PCA}
sox17gfp <- RunPCA (sox17gfp, verbose = FALSE)
```

## Non-linear dimensional reduction

```{r UMAP}
sox17gfp <- RunUMAP(sox17gfp, dims = 1:30, verbose = FALSE)
sox17gfp <- FindNeighbors(sox17gfp, dims = 1:30, verbose = FALSE)
sox17gfp <- FindClusters(sox17gfp, verbose = FALSE)

#Figure_S2A
DimPlot(sox17gfp, label = TRUE, pt.size = 1) + NoLegend() + labs(title = "GFP-expressing cells")
```

## Marker expression analysis

```{r Sox17 GFP}
#Figure_S2C
FeaturePlot(sox17gfp, features = "Sox17", pt.size = 0.5)

FeaturePlot(sox17gfp, features = "GFP", pt.size = 0.5)
```

From the violin plot of number of gene expressed, the sample split into 2 main populations, one with high and the other with low RNA content. 

```{r nFeature_RNA}
#Figure_S2C
FeaturePlot(sox17gfp, features = "nFeature_RNA", pt.size = 0.5)
```

Known markers of three main Sox17-expressing lineages

```{r Epcam Pecam1 Hba}
#Figure_S2C
#Vasculature
FeaturePlot(sox17gfp, features = "Pecam1", pt.size = 0.5)
#Endoderm
FeaturePlot(sox17gfp, features = "Epcam", pt.size = 0.5)
#Erythrocyte-related
FeaturePlot(sox17gfp, features = "Hba-a1", pt.size = 0.5)
```

Since not all clusters are positive for either Pecam1 or Epcam, differential expression analysis was performed. This analysis is just to give a general idea of what possible cell types might be present. A definite analysis of marker expression will be performed later after predicted doublets are filtered out. 

```{r cluster6_10_markers}
cluster6_10.markers <- FindMarkers(sox17gfp, ident.1 = c(6,10), min.pct = 0.25, only.pos = TRUE)
View(cluster6_10.markers)
```

GO enrichment analysis (using Enrichr by Maayan's lab) of found differentially expressed genes indicated that cluster 6 and 10 express genes related to hemoglobin and heme biosynthetic process. 

```{r cluster0_4}
cluster0_4.markers <- FindMarkers(sox17gfp, ident.1 = c(0,4), min.pct = 0.25, only.pos = TRUE)
View(cluster0_4.markers)
```

GO enrichment analysis of differentially expressed genes indicated that cluster 0 and 5 express genes related to RNA processing... Identity of cluster 0 and 4 (low RNA content) remains unclear. 

```{r}
cluster3_5_11.markers <- FindMarkers(sox17gfp, ident.1 = c(3,5,11), min.pct = 0.25, only.pos = TRUE)
View(cluster3_5_11.markers)
```

Clusters 3,4 and 11 contains cells that are enriched for Crabp2, some neural-related genes such as Sox2, Tubb3 etc. but the % of cells expressing these genes doesn't seem significant. The identiy of these clusters remain unclear. 

For record, marker genes of all clusters were analyzed

```{r}
sox17gfp.markers <- FindAllMarkers(object = sox17gfp, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
top20_sox17gfp.markers <- sox17gfp.markers %>% group_by(cluster) %>% top_n(n = 20, wt = avg_log2FC)
```

Visualize top gene markers of every cluster

```{r}
sox17gfp.markers %>% group_by(cluster) %>% top_n(n = 5, wt = avg_log2FC) -> top5

#Figure_S2D
DoHeatmap(object = sox17gfp, features = top5$gene, size = 3) + NoLegend() + theme(text = element_text(size = 15))
```

FeaturePlot for some interesting markers

```{r}
#Figure_S2C
FeaturePlot(sox17gfp, features = "Snhg9", pt.size = 0.5)
FeaturePlot(sox17gfp, features = "Crabp2", pt.size = 0.5)
FeaturePlot(sox17gfp, features = "Actb", pt.size = 0.5)
```

Since most of the low RNA content cells have unclear identity, we decided to remove the low RNA content population from further analysis. The dataset, therefore, was reanalyzed from the beginning with the min threshold for nFeature_RNA changed to 3000. 

```{r}
sox17gfp <- CreateSeuratObject(counts=sox17gfp.data, project = "gfp", min.cells = 3, min.features = 200)

sox17gfp[["percent.mt"]] <- PercentageFeatureSet(sox17gfp, pattern = "^mt-")

sox17gfp <- subset(x = sox17gfp, subset = nFeature_RNA > 3000 & nFeature_RNA < 8000 & percent.mt < 5)

#Figure_S3B
VlnPlot(sox17gfp, features = c("nFeature_RNA","nCount_RNA", "percent.mt"), ncol = 3)

plot5_gfp <- FeatureScatter(sox17gfp, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot6_gfp <- FeatureScatter(sox17gfp, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot5_gfp + plot6_gfp

sox17gfp <- SCTransform(sox17gfp, method = "glmGamPoi", vars.to.regress = "percent.mt", verbose = FALSE)

sox17gfp <- RunPCA (sox17gfp, verbose = FALSE)

sox17gfp <- RunUMAP(sox17gfp, dims = 1:30, verbose = FALSE)
sox17gfp <- FindNeighbors(sox17gfp, dims = 1:30, verbose = FALSE)
sox17gfp <- FindClusters(sox17gfp, verbose = FALSE)

#Figure_S3A
DimPlot(sox17gfp, label = TRUE, pt.size = 0.5) + NoLegend() + labs(title = "GFP-expressing cells")
```

```{r save object}
saveRDS(sox17gfp, file = "output/sox17gfp.rds")
```

# Doublet prediction

Doublet prediction is performed using DoubletFinder package.

```{r installation}
#remotes::install_github("chris-mcginnis-ucsf/DoubletFinder")
#install.packages("ROCR")
#install.packages("KernSmooth")
```

```{r setup}
library(DoubletFinder)
library(ROCR)
library(KernSmooth)
```

```{r Identify pK}
sweep.res.list_sox17gfp <- paramSweep_v3(sox17gfp, PCs = 1:30, sct=TRUE)

sweep.stats_sox17gfp <- summarizeSweep(sweep.res.list_sox17gfp, GT=FALSE)

bcmvn_sox17gfp <- find.pK(sweep.stats_sox17gfp)
```

```{r visualize pK}
sox17gfp_pK=as.numeric(as.character(bcmvn_sox17gfp$pK))
sox17gfp_BCmetric=bcmvn_sox17gfp$BCmetric
sox17gfp_pK_choose = sox17gfp_pK[which(sox17gfp_BCmetric %in% max(sox17gfp_BCmetric))]

par(mar=c(5,4,4,8)+1,cex.main=1.2,font.main=2)
plot(x = sox17gfp_pK, y = sox17gfp_BCmetric, pch = 16,type="b",
     col = "cyan3",lty=1)
abline(v=sox17gfp_pK_choose,lwd=2,col='red',lty=2)
title("The BCmvn distributions")
text(sox17gfp_pK_choose,max(sox17gfp_BCmetric),as.character(sox17gfp_pK_choose),pos = 4,col = "red")
```

```{r estimate homotypic doublet proportion}
sox17gfp_annotations <- sox17gfp@meta.data$seurat_clusters
sox17gfp_homotypic.prop <- modelHomotypic(sox17gfp_annotations)       

## Assuming 7.5% doublet formation rate
sox17gfp_nExp_poi <- round(0.075*4205)  

sox17gfp_nExp_poi.adj <- round(sox17gfp_nExp_poi*(1-sox17gfp_homotypic.prop))
```

```{r Run DoubletFinder}

sox17gfp <- doubletFinder_v3(sox17gfp, PCs=1:30, pN=0.20, pK=0.26, nExp = sox17gfp_nExp_poi, reuse.pANN = FALSE , sct = TRUE)

head(sox17gfp[[]])

sox17gfp <- doubletFinder_v3(sox17gfp, PCs = 1:30, pN = 0.20, pK = 0.26, nExp = sox17gfp_nExp_poi.adj, reuse.pANN = FALSE, sct = TRUE)
```

```{r visualize doublets}
#Figure_S3C
DimPlot(sox17gfp, pt.size = 0.5, label = F, label.size = 5, reduction = "umap", group.by = "DF.classifications_0.2_0.26_281") + labs(title = "GFP-expressing doublets")
```

```{r number of doublets}
sum(sox17gfp$DF.classifications_0.2_0.26_281 == "Doublet")
```

```{r subset to filter out doublets}
sox17gfp_nodoublet <- subset (sox17gfp, subset = (DF.classifications_0.2_0.26_281 == "Singlet"))
```

```{r save object}
saveRDS(sox17gfp_nodoublet, file = "output/sox17gfp_nodoublet.rds")
```

# Differentially expressed features (cluster biomarkers)

```{r positive markers of every cluster}
sox17gfp_nodoublet.markers <- FindAllMarkers(sox17gfp_nodoublet, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
sox17gfp_nodoublet.markers %>%
    group_by(cluster) %>%
    top_n(n = 5, wt = avg_log2FC) -> sox17gfp_nodoublet.markers.top5
write.csv(sox17gfp_nodoublet.markers.top5, file = "sox17gfp_nodoublet_marker_top5.csv")
```

```{r visualize markers}
#Figure_S3D
#Vasculature
FeaturePlot(sox17gfp_nodoublet, features = "Pecam1", pt.size = 0.5)
#Endoderm
FeaturePlot(sox17gfp_nodoublet, features = "Epcam", pt.size = 0.5)
#Erythrocyte-related
FeaturePlot(sox17gfp_nodoublet, features = "Hba-a1", pt.size = 0.5)
```

