---
title: "Analysis of Sox17-expressing lineages - Tdtomato dataset"
output:
  html_document:
    df_print: paged
---

This notebook is used to analyze a single-cell RNA sequencing dataset of Sox17-expressing lineages (also known as the TdTomato dataset in the manuscript). Sorted viable TdTomato+ cells at E9.0-9.5 from Sox17GFPCre heterozygous males breeding with R26.LSL.TdTomato homozygous females were sequenced at sequencing service core,VANTAGE(Vanderbilt Technologies for Advanced Genomics). To confirm the identity of sorted TdTomato+cells, a customized genome build was generated by adding in Tdtomato sequence to 10x Genomics reference (References - 2020-A, mm10) followed by standard alignment, UMI counting and barcode filtering using CellRanger 6.0.2

***

This analysis is performed with Seurat4 package. If Seurat is not installed or if certain version of Seurat is required, refer to satijalab.org for installation instructions.

***

# Setting up

```{r setup environment, include=FALSE}
all_times <- list() 
knitr::knit_hooks$set(time_it = local({
  now <- NULL
  function(before, options) {
    if(before){
      # report the current time before each chunk
      now <<- Sys.time()
    } else {
      # calculate the time difference after a chunk
      res <- difftime(Sys.time(), now, units = "secs")
      # store the reported time it takes to run each chunk to all_time object
      all_times[[options$label]] <<- res
    }
  }
}))
knitr::opts_chunk$set(
  # to reformat all R source code in several aspects such as adding spaces around most operators, indenting the code properly and replacing the assignment operator = into <-
  tidy = TRUE,
  # to set cutoff for the width of output text
  tidy.opts = list (width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  time_it = TRUE
)
```

```{r initiate library}
library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)
```

```{r load data}
sox17tdtomato.data <- Read10X(data.dir = "data/201214/filtered_feature_bc_matrix/")
```

```{r setup Seurat object}
sox17tdtomato <- CreateSeuratObject(counts=sox17tdtomato.data, project = "tdtom", min.cells = 3, min.features = 200)

sox17tdtomato
```

# Pre-processing

## Quality control

```{r percentage mitochondrial gene}

sox17tdtomato[["percent.mt"]] <- PercentageFeatureSet(sox17tdtomato, pattern = "^mt-")
```

```{r visualize QC metrics - before filter}

VlnPlot(sox17tdtomato, features = c("nFeature_RNA","nCount_RNA", "percent.mt"), ncol = 3)

plot1_tdtomato <- FeatureScatter(sox17tdtomato, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2_tdtomato <- FeatureScatter(sox17tdtomato, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1_tdtomato + plot2_tdtomato
```

Filtering criteria is similar to the analysis of GFP dataset

```{r filter low-qual cells}
# Filter out cells that have unique feature counts over 8,000 or less than 3000 or have more than 5% mitochondrial counts
sox17tdtomato <- subset(x = sox17tdtomato, subset = nFeature_RNA > 3000 & nFeature_RNA < 8000 & percent.mt < 5)
```

```{r visualize QC metrics after}
#Figure_S3F
VlnPlot(sox17tdtomato, features = c("nFeature_RNA","nCount_RNA", "percent.mt"), ncol = 3)

plot3_tdtomato <- FeatureScatter(sox17tdtomato, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot4_tdtomato <- FeatureScatter(sox17tdtomato, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot3_tdtomato + plot4_tdtomato
```

## Normalization

Similar to what was done with GPF dataset

```{r glmGamPoi}
#glmGamPoi package can also be used to substantially improves the speed of the learning procedure. It can be invoked by specifying method="glmGamPoi"

#if(!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
#BiocManager::install("glmGamPoi")
```

```{r sctransform}
sox17tdtomato <- SCTransform(sox17tdtomato, method = "glmGamPoi", vars.to.regress = "percent.mt", verbose = FALSE)
```

# Dimensional reduction

## Linear dimentional reduction PCA

```{r PCA}
sox17tdtomato <- RunPCA (sox17tdtomato, verbose = FALSE)
```

## Non-linear dimensional reduction

```{r UMAP}
sox17tdtomato <- RunUMAP(sox17tdtomato, dims = 1:30, verbose = FALSE)
sox17tdtomato <- FindNeighbors(sox17tdtomato, dims = 1:30, verbose = FALSE)
sox17tdtomato <- FindClusters(sox17tdtomato, verbose = FALSE)

#Figure_S3E
DimPlot(sox17tdtomato, label = TRUE) + NoLegend() + labs(title = "TdTomato-expressing cells")
```

## Marker expression analysis

```{r Sox17 GFP TdTomato}
FeaturePlot(sox17tdtomato, features = "Sox17", pt.size = 0.5)
FeaturePlot(sox17tdtomato, features = "Gfp", pt.size = 0.5)
FeaturePlot(sox17tdtomato, features = "Tdtomato", pt.size = 0.5)
```

Known markers of three main Sox17-expressing lineages

```{r Epcam Pecam1 Hba}
#Figure_S3H
#Vasculature
FeaturePlot(sox17tdtomato, features = "Pecam1", pt.size = 0.5)
#Endoderm
FeaturePlot(sox17tdtomato, features = "Epcam", pt.size = 0.5)
#Erythrocyte-related
FeaturePlot(sox17tdtomato, features = "Hba-a1", pt.size = 0.5)
```

## Save object (optional)

```{r save object}
saveRDS(sox17tdtomato, file = "output/sox17tdtomato.rds")
```

# Doublet prediction

```{r installation}
#remotes::install_github("chris-mcginnis-ucsf/DoubletFinder")
#install.packages("ROCR")
#install.packages("KernSmooth")
```

```{r setup}
library(DoubletFinder)
library(ROCR)
library(KernSmooth)
```

```{r Identify pK}

sweep.res.list_sox17tdtomato <- paramSweep_v3(sox17tdtomato, PCs = 1:30, sct=TRUE)

sweep.stats_sox17tdtomato <- summarizeSweep(sweep.res.list_sox17tdtomato, GT=FALSE)

bcmvn_sox17tdtomato <- find.pK(sweep.stats_sox17tdtomato)

```

```{r visualize pK}
sox17tdtomato_pK=as.numeric(as.character(bcmvn_sox17tdtomato$pK))
sox17tdtomato_BCmetric=bcmvn_sox17tdtomato$BCmetric
sox17tdtomato_pK_choose = sox17tdtomato_pK[which(sox17tdtomato_BCmetric %in% max(sox17tdtomato_BCmetric))]

par(mar=c(5,4,4,8)+1,cex.main=1.2,font.main=2)
plot(x = sox17tdtomato_pK, y = sox17tdtomato_BCmetric, pch = 16,type="b",
     col = "cyan3",lty=1)
abline(v=sox17tdtomato_pK_choose,lwd=2,col='red',lty=2)
title("The BCmvn distributions")
text(sox17tdtomato_pK_choose,max(sox17tdtomato_BCmetric),as.character(sox17tdtomato_pK_choose),pos = 4,col = "red")
```


```{r estimate homotypic doublet proportion}
sox17tdtomato_annotations <- sox17tdtomato@meta.data$seurat_clusters
sox17tdtomato_homotypic.prop <- modelHomotypic(sox17tdtomato_annotations)   

## Assuming 7.5% doublet formation rate

sox17tdtomato_nExp_poi <- round(0.075*3221)  

sox17tdtomato_nExp_poi.adj <- round(sox17tdtomato_nExp_poi*(1-sox17tdtomato_homotypic.prop))
```


```{r Run DoubletFinder}

sox17tdtomato <- doubletFinder_v3(sox17tdtomato, PCs=1:30, pN=0.20, pK=0.03, nExp = sox17tdtomato_nExp_poi, reuse.pANN = FALSE , sct = TRUE)

head(sox17tdtomato[[]])

sox17tdtomato <- doubletFinder_v3(sox17tdtomato, PCs = 1:30, pN = 0.20, pK = 0.03, nExp = sox17tdtomato_nExp_poi.adj, reuse.pANN = FALSE , sct = TRUE)

```


```{r visualize doublets}
#Figure_S3G
DimPlot(sox17tdtomato, pt.size = 0.5, label = F, label.size = 5, reduction = "umap", group.by = "DF.classifications_0.2_0.03_211") + labs(title = "TdTomato-expressing doublets")
```

```{r number of doublets}
sum(sox17tdtomato$DF.classifications_0.2_0.03_211 == "Doublet")
```

```{r subset to filter out doublets}
sox17tdtomato_nodoublet <- subset (sox17tdtomato, subset = (DF.classifications_0.2_0.03_211 == "Singlet"))
```

```{r save object}
saveRDS(sox17tdtomato_nodoublet, file = "output/sox17tdtomato_nodoublet.rds")
```

# Differentially expressed features (cluster biomarkers)

```{r positive markers of every cluster}
sox17tdtomato_nodoublet.markers <- FindAllMarkers(sox17tdtomato_nodoublet, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
sox17tdtomato_nodoublet.markers %>%
    group_by(cluster) %>%
    top_n(n = 5, wt = avg_log2FC) -> sox17tdtomato_nodoublet.markers.top5
write.csv(sox17tdtomato_nodoublet.markers.top5, file = "sox17tdtomato_nodoublet_marker_top5.csv")
```
